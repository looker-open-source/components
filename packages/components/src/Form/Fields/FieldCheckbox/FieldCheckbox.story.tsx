/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React from 'react'
import { Story } from '@storybook/react/types-6-0'
import { FieldCheckbox, FieldCheckboxProps } from './FieldCheckbox'

const Template: Story<FieldCheckboxProps> = (args) => (
  <FieldCheckbox {...args} />
)

export const Basic = Template.bind({})
Basic.args = {
  id: 'id',
  label: 'Example Field',
  name: 'thumbsUp',
}

export const DetailDescription = Template.bind({})
DetailDescription.args = {
  ...Basic.args,
  description: 'describe something here.',
  detail: '4/20',
}

export const Checked = Template.bind({})
Checked.args = {
  ...Basic.args,
  checked: true,
}

export const DefaultChecked = Template.bind({})
DefaultChecked.args = {
  ...Basic.args,
  defaultChecked: true,
}

export const Disabled = Template.bind({})
Disabled.args = {
  ...Basic.args,
  disabled: true,
}

export const DisabledChecked = Template.bind({})
DisabledChecked.args = {
  ...Checked.args,
  disabled: true,
}

export const ReadOnly = Template.bind({})
ReadOnly.args = {
  ...Basic.args,
  readOnly: true,
}

export const Error = Template.bind({})
Error.args = {
  ...Basic.args,
  validationMessage: { message: 'This is an error', type: 'error' },
}

export const DetailDescriptionError = Template.bind({})
DetailDescriptionError.args = {
  ...DetailDescription.args,
  validationMessage: { message: 'This is an error', type: 'error' },
}

export default {
  component: FieldCheckbox,
  title: 'FieldCheckbox',
}
