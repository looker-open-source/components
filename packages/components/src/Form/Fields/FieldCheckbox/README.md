# FieldCheckbox

The `<FieldCheckbox />` component is composed of a `<Checkbox />` component and a `<Label />` component. Using `<FieldCheckbox />` allows for rendering validation messages.

```tsx
function Basic(props: FieldCheckboxProps) {
  return (
    <FieldCheckbox id="id" label="Example Field" name="thumbsUp" {...props} />
  );
}
```

### FieldCheckbox Validation

The `<FieldCheckbox />` components accepts `validationMessage` as a property. This is an object with properties `type` and `message`.

See `ValidationMessage` [here](?path=/docs/docs-validationmessage--docs)

If you have multiple fields in need of validation, it is recommended that you wrap them in a `<Form />` component and pass to it a dictionary with key as field name and value as a `ValidationMessage`.

The `alignValidationMessage` can also be used to adjust where the validation message appears in relation to the input field. By default, it will render to the right, which is the best practice in most cases.

**_Note: In the current implementation, any `ValidationMessage` set in a `<FieldCheckbox />` component will be overridden by `ValidationMessages` set in the parent `<Form />` component. Because of this, it is highly recommended that you only ever directly set `ValidationMessage` if a `<FieldCheckbox />` is outside a `<Form />`._**

```tsx
function Validation() {
  return (
    <FieldCheckbox
      id="id"
      label="Validation error"
      name="thumbsUp"
      validationMessage={{ message: 'This is an error', type: 'error' }}
    />
  );
}
```
