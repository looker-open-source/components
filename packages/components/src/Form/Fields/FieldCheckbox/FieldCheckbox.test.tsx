/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import 'jest-styled-components'
import React from 'react'
import {
  assertSnapshot,
  mountWithTheme,
  renderWithTheme,
} from '@looker/components-test-utils'
import { FieldCheckbox } from './FieldCheckbox'

test('A FieldCheckbox', () => {
  assertSnapshot(
    <FieldCheckbox id="FieldCheckboxID" label="👍" name="thumbsUp" />
  )
})

test('A FieldCheckbox with checked value', () => {
  assertSnapshot(
    <FieldCheckbox
      defaultChecked
      id="FieldCheckboxID"
      label="👍"
      name="thumbsUp"
    />
  )
})

test('A required FieldCheckbox', () => {
  const wrapper = mountWithTheme(
    <FieldCheckbox id="FieldCheckboxID" label="👍" name="thumbsUp" required />
  )
  expect(wrapper.text()).toMatch(`👍 required`)
})

test('A disabled FieldCheckbox', () => {
  const wrapper = mountWithTheme(
    <FieldCheckbox disabled id="FieldCheckboxID" label="👍" name="thumbsUp" />
  )
  wrapper.find('input').html().includes('disabled=""')
})

test('A FieldCheckbox with error has proper aria setup', () => {
  const errorMessage = 'This is an error'

  const { container, getByDisplayValue } = renderWithTheme(
    <FieldCheckbox
      id="test"
      defaultValue="example"
      validationMessage={{ message: errorMessage, type: 'error' }}
    />
  )

  const input = getByDisplayValue('example')
  const id = input.getAttribute('aria-describedby')
  expect(id).toBeDefined()

  const describedBy = container.querySelector(`#${id}`)
  expect(describedBy).toHaveTextContent(errorMessage)
})
