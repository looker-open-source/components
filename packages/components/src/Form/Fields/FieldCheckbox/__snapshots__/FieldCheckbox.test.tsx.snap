// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A FieldCheckbox 1`] = `
.c6 {
  position: relative;
  width: 100%;
  height: 100%;
  background: currentColor;
  border: solid 1px #DEE1E5;
  border-radius: 0.125rem;
  color: #FFFFFF;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4 {
  height: 1rem;
  position: relative;
  width: 1rem;
}

.c4 input {
  height: 100%;
  opacity: 0;
  position: absolute;
  width: 100%;
  z-index: 1;
}

.c4 input:checked + .c5 {
  background-color: #6C43E0;
  border-color: #6C43E0;
}

.c4 input:not(:checked) + .c5 {
  color: #FFFFFF;
}

.c4 input:focus + .c5 {
  border-color: #9785F2;
  box-shadow: 0 0 0 2px #E8E5FF;
  outline: none;
}

.c4 input:disabled + .c5 {
  background: #F5F6F7;
  border-color: #DEE1E5;
  color: #939BA5;
}

.c4 input:disabled:not(:checked) + .c5 {
  color: transparent;
}

.c2 {
  color: #343C42;
  font-size: 0.75rem;
  font-weight: 600;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: grid;
  grid-template-areas: 'input label detail' '. messages messages';
  grid-template-columns: repeat(2,max-content) 1fr;
  line-height: 1.25rem;
}

.c0 .c3 {
  grid-area: input;
}

.c0 .c1 {
  font-size: 0.875rem;
  font-weight: normal;
  grid-area: label;
  padding-left: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .c8 {
  grid-area: detail;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  margin-left: 0.25rem;
}

.c0 .c7 {
  grid-area: messages;
  padding-left: 0.75rem;
}

<label
  className="c0 "
  htmlFor="FieldCheckboxID"
>
  <span
    className="c1 c2"
    color="text2"
    fontSize="xsmall"
    fontWeight="semiBold"
  >
    üëç
  </span>
  <div
    className="c3 "
  >
    <div
      className="c4 "
    >
      <input
        aria-describedby="FieldCheckboxID-describedby"
        checked={false}
        id="FieldCheckboxID"
        name="thumbsUp"
        onChange={[Function]}
        onClick={[Function]}
        type="checkbox"
      />
      <div
        className="c5 c6"
      >
        <svg
          fill="none"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 8L7 11L12 4"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    className="c7 "
    id="FieldCheckboxID-describedby"
  />
</label>
`;

exports[`A FieldCheckbox with checked value 1`] = `
.c6 {
  position: relative;
  width: 100%;
  height: 100%;
  background: currentColor;
  border: solid 1px #DEE1E5;
  border-radius: 0.125rem;
  color: #FFFFFF;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4 {
  height: 1rem;
  position: relative;
  width: 1rem;
}

.c4 input {
  height: 100%;
  opacity: 0;
  position: absolute;
  width: 100%;
  z-index: 1;
}

.c4 input:checked + .c5 {
  background-color: #6C43E0;
  border-color: #6C43E0;
}

.c4 input:not(:checked) + .c5 {
  color: #FFFFFF;
}

.c4 input:focus + .c5 {
  border-color: #9785F2;
  box-shadow: 0 0 0 2px #E8E5FF;
  outline: none;
}

.c4 input:disabled + .c5 {
  background: #F5F6F7;
  border-color: #DEE1E5;
  color: #939BA5;
}

.c4 input:disabled:not(:checked) + .c5 {
  color: transparent;
}

.c2 {
  color: #343C42;
  font-size: 0.75rem;
  font-weight: 600;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: grid;
  grid-template-areas: 'input label detail' '. messages messages';
  grid-template-columns: repeat(2,max-content) 1fr;
  line-height: 1.25rem;
}

.c0 .c3 {
  grid-area: input;
}

.c0 .c1 {
  font-size: 0.875rem;
  font-weight: normal;
  grid-area: label;
  padding-left: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .c8 {
  grid-area: detail;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  margin-left: 0.25rem;
}

.c0 .c7 {
  grid-area: messages;
  padding-left: 0.75rem;
}

<label
  className="c0 "
  htmlFor="FieldCheckboxID"
>
  <span
    className="c1 c2"
    color="text2"
    fontSize="xsmall"
    fontWeight="semiBold"
  >
    üëç
  </span>
  <div
    className="c3 "
  >
    <div
      className="c4 "
    >
      <input
        aria-describedby="FieldCheckboxID-describedby"
        checked={true}
        id="FieldCheckboxID"
        name="thumbsUp"
        onChange={[Function]}
        onClick={[Function]}
        type="checkbox"
      />
      <div
        className="c5 c6"
      >
        <svg
          fill="none"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 8L7 11L12 4"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    className="c7 "
    id="FieldCheckboxID-describedby"
  />
</label>
`;
